lib/jobs/ucs-discovery.js

```diff
@@ -0,0 +1,169 @@
+                                   context,
+                                   taskId);
+
+        assert.object(this.options);
+        assert.string(this.options.uri);
+        var parse = urlParse(this.options.uri);
+        var protocol = parse.protocol.replace(':','').trim();
+        this.settings = {
+            uri: parse.href,
+            host: parse.host.split(':')[0],
+            root: parse.pathname,
+            port: parse.port,
+            protocol: protocol,
+            username: this.options.username,
+            password: this.options.password,
+            ucs: this.options.ucs,
+            verifySSL: this.options.verifySSL || false
+        };
+        this.ucs = new UcsTool();
+        this.ucs.settings = this.settings;
+    }
+    
+    util.inherits(UcsDiscoveryJob, BaseJob);
+
+    /**
+     * @memberOf UcsDiscoveryJob
+     */
+    UcsDiscoveryJob.prototype._run = function() {
+        var self = this;
+        
+        return self.getRoot()
+            .then(function(root) {
+                return [ root, self.createRackmounts(root) ];
+            })
+            .then(function() {
+                self._done();
+            })
+            .catch(function(err) {
+                self._done(err);
+            });
+    };
+
+    /**
+     * @function logIn
+     * @description Log into the UCS microservice
+     */
+    UcsDiscoveryJob.prototype.logIn = function () {
+        var username = this.settings.username;
+        var password = this.settings.password;
+        var host = this.settings.ucs;
+        var url= "/login?host=" + host+ "&user="+ username + "&password=" + password;
+        var d ={};
+        d.username=this.settings.username;
+        d.password=this.settings.password;
+        return this.ucs.clientRequest(url)
+            .then(function(response) {
+                return response.body;
+            });
+    };
+
+
+    /**
+     * @function getRoot
+     * @description return all the elments in the Cisco UCS that RackHD cares about
+     */
+    UcsDiscoveryJob.prototype.getRoot = function () {
+        var username = this.settings.username;
+        var password = this.settings.password;
+        var ucs = this.settings.ucs;
+        var baseurl= this.settings.protocol+"://"+this.settings.host+ ":"+this.settings.port;
+        var url= baseurl = "/sys?host=" + ucs+ "&user="+ username + "&password=" + password;/* jshint ignore:line */
+
+        return this.ucs.clientRequest(url)
+            .then(function(response) {
+                return response.body;
+            });
+    };
+
+    /**
+     * @function createRackmounts
+     * @description discovers all the rackmount servers in the Cisco UCS
+     */
+    UcsDiscoveryJob.prototype.createRackmounts = function (root) {
+        var username = this.settings.username;
+        var password = this.settings.password;
+        var ucs = this.settings.ucs;
+
+        if (!_.has(root, 'Servers')) {
```

Comment:
@tannoa2 I recommend wrapping this code in a `Promise.try` to avoid throwing a synchronous error.  This ensures that this function will always return a Promise.